src/Control/Optics/Linear/Internal.hs

```diff
@@ -6,31 +6,442 @@
 {-# LANGUAGE RankNTypes #-}
 {-# LANGUAGE TypeOperators #-}
 
+-- | This module provides linear optics.
+--
+-- == Overview
+--
+-- /Some familiarity with optics is needed to understand linear optics./
+-- /Please go through the (hopefully friendly!) background material section/
+-- /if you are unfamiliar with lenses, prisms, traversals or isomorphisms./
+--
+-- This documentation provides
+--
+--  * Some bird's eye views on (1) what linear optics are, (2) how to think
+--  about them, and (3) an overview of the types
+--  * Examples of using these optics
+--
+-- == Background Material
+--
+--  * [A great intro-to-lenses talk by Simon Peyton Jones](https://skillsmatter.com/skillscasts/4251-lenses-compositional-data-access-and-manipulation)
+--  * [A nice introductory blog post](https://tech.fpcomplete.com/haskell/tutorial/lens)
+--  * [A friendly introduction to prisms and isos](https://www.schoolofhaskell.com/school/to-infinity-and-beyond/pick-of-the-week/a-little-lens-starter-tutorial)
+--  * [The wiki of the @lens@ package](https://github.com/ekmett/lens/wiki)
+--  that contains some nice examples
+--
+-- == Conceptualizing and using optics
+--
+-- === What are /linear/ optics?
+--
+-- Optics can be conceptualized as a first class object with which you can
+-- view and map functions over sub-structure(s) within a larger structure.
+--
+-- __ /Linear/ optics are optics where the \"viewing\" and \"mapping\" are__
+-- __done with linear functions (and any corresponding structures hold values__
+-- __linearly, i.e., with constructors that use linear arrows).__
+--
+-- In types: a (linear) optic of type @Optic a b s t@ is a way of viewing the
+-- sub-structure(s) of type @a@ in the structure of type @s@ and mapping a
+-- function from an @a@ to a @b@ on those sub-structures in @s@ which change an
+-- @s@ to a @t@. The non-polymorphic version of the optic is specialized
+-- to the types @Optic a a s s@ and is usually defined with a tick mark,
+-- e.g., the non-polymorphic @Lens@ is @Lens'@.
+--
+-- There are four basic optics: traversals, lenses, prisms and isomorphisms.
+--
+-- ==== Sub-typing diagram of optics
+--
+-- \[ \texttt{Traversal} \]
+-- \[ \Huge \nearrow ~~~~~ \nwarrow \]
+-- \[ \texttt{Lens}\hspace{6em}\texttt{Prism} \]
+-- \[ \Huge \nwarrow ~~~~~ \nearrow \]
+-- \[ \texttt{Iso} \]
+--
+-- In the diagram above, the arrows @X --> Y@ mean any of the following
+-- equivalent things:
+--
+--  * X is a specialization of Y
+--  * X is a strict subset of Y
+--  * You can (basically) implement @f :: X -> Y@ with @f = id@
+--  but you can't implement @f :: Y -> X@.
+--
+-- === How can each optic be used? How do I think about them?
+--
+-- * A @Traversal a b s t@ is roughly equivalent to a
+--  @s \#-> (Tree a, Tree b \#-> t)@.
```

Comment:
Why is there a `Tree` here?

Reply:
I'm cutting this section almost entirely. Not ready to write it yet.